File: BOJ/5014.cpp
#include <iostream>
#include <string>
#include <queue>
#include <algorithm>
#define MAX 1000001
using namespace std;

int f, s, g, u, d;
int visited[MAX];

void bfs(){
    queue<int> q;
    q.push(s);
    visited[s] = 0;
    
    while(!q.empty()){
        int x = q.front();
        q.pop();
        
        if(x == g){
            cout << visited[x] << '\n';
            return;
        }
        
        for(int i=0; i<2; i++){
            int dx;
            if(i==0) dx = x + u;
            else     dx = x - d;
            
            if(dx > f || dx <= 0) continue;
            if(visited[dx] >= 0) continue;
            visited[dx] = visited[x] + 1;
            q.push(dx);
        }
    }
    cout << "use the stairs\n";
}

int main(){
    ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    cin >> f >> s >> g >> u >> d;
    for(int i=1; i<=f; i++) visited[i] = -1;
    bfs();
    return 0;
}

File: BOJ/Swift/5014.swift
import Foundation

/*
 F층 빌딩. 갈 곳 G층. 지금있는 층 S층
 U버튼: 위로 U층, D버튼: 아래로 D층
 G층에 가려면 버튼을 적어도 몇 번 눌러야?
 */

let input = readLine()!.split(separator: " ").map{ Int(String($0))! }
let f = input[0]
let s = input[1]
let g = input[2]
let u = input[3]
let d = input[4]
var visited = [Int](repeating: -1, count: f+1)

func BFS() {
    var queue = [s]
    var front = 0
    visited[s] = 0
    
    while queue.count > front {
        let x = queue[front]
        front += 1
        
        if g == x {
            print(visited[x])
            return
        }
        
        let dir = [u, -d]
        for i in 0..<2 {
            let nx = x + dir[i]
            
            guard (1...f) ~= nx else { continue }
            guard visited[nx] == -1 else { continue }
            visited[nx] = visited[x] + 1
            queue.append(nx)
        }
    }
    print("use the stairs")
}

BFS()
